src/translate_c.zig

```diff
@@ -1787,52 +1796,74 @@ fn transStringLiteral(
     }
 }
 
-/// Parse the size of an array back out from an ast Node.
-fn zigArraySize(c: *Context, node: Node) TransError!usize {
-    if (node.castTag(.array_type)) |array| {
-        return array.data.len;
-    }
-    return error.UnsupportedTranslation;
+fn zigArraySize(array_node: Node) usize {
+    if (array_node.castTag(.array_type)) |arr| return arr.data.len;
+    if (array_node.castTag(.null_sentinel_array_type)) |arr| return arr.data.len;
+    unreachable;
 }
 
-/// Translate a string literal to an array of integers. Used when an
-/// array is initialized from a string literal. `array_size` is the
-/// size of the array being initialized. If the string literal is larger
-/// than the array, truncate the string. If the array is larger than the
-/// string literal, pad the array with 0's
-fn transStringLiteralAsArray(
+fn zigArrayElemType(array_node: Node) Node {
+    if (array_node.castTag(.array_type)) |arr| return arr.data.elem_type;
+    if (array_node.castTag(.null_sentinel_array_type)) |arr| return arr.data.elem_type;
+    unreachable;
+}
+
+/// Translate a string literal that is initializing an array. In general narrow string
+/// literals become `"<string>"[0..<size>].*` (unless the length is zero)
+/// Wide string literals become an array of integers. zero-fillers pad out the array to
+/// the appropriate length, if necessary.
+fn transStringLiteralInitializer(
     c: *Context,
     scope: *Scope,
     stmt: *const clang.StringLiteral,
-    array_size: usize,
+    array_type: Node,
 ) TransError!Node {
-    if (array_size == 0) return error.UnsupportedType;
+    assert(array_type.tag() == .array_type or array_type.tag() == .null_sentinel_array_type);
+
+    const is_narrow = stmt.getKind() == .Ascii or stmt.getKind() == .UTF8;
 
     const str_length = stmt.getLength();
+    const array_size = zigArraySize(array_type);
+    const elem_type = zigArrayElemType(array_type);
+
+    if (array_size == 0) return Tag.empty_array.create(c.arena, elem_type);
+
+    const num_inits = math.min(str_length, array_size);
+    const init_node = if (num_inits > 0) blk: {
+        if (is_narrow) {
+            // "string literal"[0..num_inits].*
+            const lit_node = try transNarrowStringLiteral(c, scope, stmt, .used);
+            const sliced = try Tag.string_slice.create(c.arena, .{ .string = lit_node, .end = num_inits });
+            break :blk try Tag.deref.create(c.arena, sliced);
```

Comment:
```suggestion
            var str = try transNarrowStringLiteral(c, scope, stmt, .used);
            if (str_length != array_size) str = try Tag.string_slice.create(c.arena, .{ .string = str, .end = num_inits });
            break :blk try Tag.deref.create(c.arena, str);
```
Looks like we should be able to avoid the slice most of the time.